src/qpx_mmc_css.cpp

```diff
@@ -0,0 +1,1909 @@
+
+  /*
+   * Prepare tables for hash reversal
+   */
+
+  /* initialize lookup tables for k[1] */
+  K1table = (unsigned char *)malloc(65536 * K1TABLEWIDTH);
+  memset(K1table, 0, 65536 * K1TABLEWIDTH);
+  if (K1table == NULL) {
+    return -1;
+  }
+
+  tmp = p_disc_key[0] ^ p_css_tab1[p_disc_key[1]];
+  for (i = 0; i < 256; i++) /* k[1] */
+  {
+    tmp2 = p_css_tab1[tmp ^ i]; /* p_css_tab1[ B[1] ]*/
+
+    for (j = 0; j < 256; j++) /* B[0] */
+    {
+      tmp3 = j ^ tmp2 ^ i; /* C[1] */
+      tmp4 =
+          K1table[K1TABLEWIDTH * (256 * j + tmp3)]; /* count of entries  here */
+      tmp4++;
+      /*
+                  if( tmp4 == K1TABLEWIDTH )
+                  {
+                      print_debug( dvdcss, "Table disaster %d", tmp4 );
+                  }
+      */
+      if (tmp4 < K1TABLEWIDTH) {
+        K1table[K1TABLEWIDTH * (256 * j + tmp3) + tmp4] = i;
+      }
+      K1table[K1TABLEWIDTH * (256 * j + tmp3)] = tmp4;
+    }
+  }
+
+  /* Initing our Really big table */
+  BigTable = (unsigned int *)malloc(16777216 * sizeof(int));
+  memset(BigTable, 0, 16777216 * sizeof(int));
+  if (BigTable == NULL) {
+    return -1;
+  }
+
+  tmp3 = 0;
+
+  printf("CSS: initializing the big table\n");
+
+  for (i = 0; i < 16777216; i++) {
+    tmp = ((i + i) & 0x1fffff0) | 0x8 | (i & 0x7);
+
+    for (j = 0; j < 5; j++) {
+      tmp2 = (((((((tmp >> 3) ^ tmp) >> 1) ^ tmp) >> 8) ^ tmp) >> 5) & 0xff;
+      tmp = (tmp << 8) | tmp2;
+      out2[j] = p_css_tab4[tmp2];
+    }
+
+    j = (out2[0] << 16) | (out2[1] << 8) | out2[4];
+    BigTable[j] = i;
+  }
+
+  /*
+   * We are done initing, now reverse hash
+   */
+  tmp5 = p_disc_key[0] ^ p_css_tab1[p_disc_key[1]];
+
+  for (nStepA = 0; nStepA < 65536; nStepA++) {
+    lfsr1a = 0x100 | (nStepA >> 8);
+    lfsr1b = nStepA & 0xff;
+
+    /* Generate 5 first output bytes from lfsr1 */
+    for (i = 0; i < 5; i++) {
+      tmp = p_css_tab2[lfsr1b] ^ p_css_tab3[lfsr1a];
+      lfsr1b = lfsr1a >> 1;
+      lfsr1a = ((lfsr1a & 1) << 8) ^ tmp;
+      out1[i] = p_css_tab4[tmp];
+    }
+
+    /* cumpute and cache some variables */
+    C[0] = nStepA >> 8;
+    C[1] = nStepA & 0xff;
+    tmp = p_disc_key[3] ^ p_css_tab1[p_disc_key[4]];
+    tmp2 = p_css_tab1[p_disc_key[0]];
+
+    /* Search through all possible B[0] */
+    for (nStepB = 0; nStepB < 256; nStepB++) {
+      /* reverse parts of the mangling cipher */
+      B[0] = nStepB;
+      k[0] = p_css_tab1[B[0]] ^ C[0];
+      B[4] = B[0] ^ k[0] ^ tmp2;
+      k[4] = B[4] ^ tmp;
+      nPossibleK1 = K1table[K1TABLEWIDTH * (256 * B[0] + C[1])];
+
+      /* Try out all possible values for k[1] */
+      for (nTry = 0; nTry < nPossibleK1; nTry++) {
+        k[1] = K1table[K1TABLEWIDTH * (256 * B[0] + C[1]) + nTry + 1];
+        B[1] = tmp5 ^ k[1];
+
+        /* reconstruct output from LFSR2 */
+        tmp3 = (0x100 + k[0] - out1[0]);
+        out2[0] = tmp3 & 0xff;
+        tmp3 = tmp3 & 0x100 ? 0x100 : 0xff;
```

Comment:
Codacy found an issue: [Clarify calculation precedence for '&' and '?'.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)